### PR TITLE
Add CGCNNFeaturizer by extracting features from a CGCNN model

### PR DESCRIPTION
**Pull request by Qi-max**
## Summary

Add CGCNNFeaturizer for generating structural features from a Crystal Graph Convolutional Neural Network (CGCNN) model.

This featurizer requires a CGCNN model that can either be:
1. from a pretrained model, currently supports models from the CGCNN repo: https://github.com/txie-93/cgcnn;
2. train a CGCNN model, based on the `X` (structures) and `y` (target) from a fresh start; 
3. similar to 2), but train a model from a `warm_start` model, which can either be a pretrained model or saved checkpoints.

This is done in the `fit.`

After obtaining a CGCNN model, we will `featurize` the structures by taking the crystal feature vector obtained after pooling as the features.

This featurizer requires installing `cgcnn` and `torch`. But as `cgcnn` is not in `PYPI`, we do not add it in the requirements. We wrap and refractor some classes and functions from the original CGCNN repo to make them work better for matminer, such as skipping the extracting structures from the CIF data as we already have `X` and `y`. The wrapper classes and functions are in`utils/cgcnn`.

* Add CGCNNFeaturizer
* Add some utils for making the training easier
* Add default initial atom features of the cgcnn repo
* Add unit tests for both CGCNNFeaturizer and cgcnn utils

## TODO (if any)

* CGCNNFeaturizer is not supported by `fit_featurize_dataframe` which assumes no `y` as input.

---
This change is [Reviewable](https://reviewable.io/reviews/hackingmaterials/matminer/345)

### CODE REVIEW

**Comment by Qi-max** [comment]
@computron I have solved the global args problem which can be done inside the CGCNNFeaturizer. So I leave the CGCNNFeaturizer in the structure featurizer, is this good? Solving this allows us to call the train and validate in the cgcnn repo.

**Comment by ardunn** [comment]
@Qi-max looks good to me, with 2 comments/suggestions.

I think before it is merged we should address the codacy issues (minus the fit parameters one).

Could you include putting y=None default kwarg in `BaseFeaturizer.fit_featurize_dataframe` so then CGCNN.fit_featurize_dataframe would be supported?

**Comment by ardunn** [comment]
@Qi-max could we also move the /tmp to the current directory the user is in? Or maybe just the matminer directory (i.e., it makes a ./models/cgcnn directory)?

**Comment by computron** [comment]
I haven't followed the /tmp directory thing, but if you are looking to just make a Scratch or Temporary , you can use monty's ScratchDir or Python's TempDir https://docs.python.org/3/library/tempfile.html

**Comment by Qi-max** [comment]
@computron  @ardunn  Thanks for the comments! I guess I can solve the Codacy issues in a later commit except for the "fit(self, X, y)" one. I also included a y=None in BaseFeaturizer.fit_featurize_dataframe. For "/tmp", the two places of using it in the previous codes are for different purposes:

- For the output_path="\tmp\CGCNNFeaturizer": This is intended to be the default output_path to store the checkpoints and models (if save_checkpoint=True or save_model=True). As using “\tmp” is not favored by Codacy, I guess I would probably just set the default to None, and if the user wants to save the checkpoints and models, they have to supply their own “meaningful" folder path? I will check this in __init__, and if the save_model or save_checkpoint is True but the output_path is None, just throw an error. Is this acceptable?

- For the sys.argv += ["\tmp"]: Indeed any string is acceptable, not necessarily "/tmp". The reason I have to add a folder name to sys.argv is because the train() and validate() functions of the CGCNN repo are embedded in its main.py, which needs command-line arguments, and a folder name designating where the cifs etc are stored is required. As now we already wrap the CIFData using a customized dataset wrapper, we do not actually have to provide a folder name. So I just added "/tmp" to hold the required place of the folder name to avoid the ArgumentError. I will use “place_holder” instead in a later commit.

The Codacy issue of fit(self, X, y) cannot be resolved as it does require a y.

**Comment by computron** [comment]
I added a bunch of comments

Overall I think this is good but it would be better if we can coordinate with the CGCNN folks (Xie Tian) so we don't reproduce so much code. Perhaps this is not possible with the way their code is written right now, but we can see if they are interested in making small changes to their code infrastructure to make this possible.

**Comment by computron** [comment]
regarding the previous comments:

>> Is this acceptable?

see my comments about combining variables, I think this will be more elegant than what you propose

>> Indeed any string is acceptable, not necessarily "/tmp"

If possible you can use "_" as the dummy argument - the single underscore usually indicates an unused or dummy variable in Python. You should also add a comment to the code clarifying that this is a dummy argument needed for the wrapper.

**Comment by Qi-max** [comment]
Hi @ardunn When I pushed the last two commits, circleci raises an error in "test_violin". Have you encountered this before?
======================================================================
FAIL: test_violin (test_plots.PlotlyFigTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/circleci/matminer/matminer/figrecipes/tests/test_plots.py", line 167, in test_violin
    self.assertTrue(vio_test == vio_true)
AssertionError: False is not true

**Comment by ardunn** [comment]
@Qi-max yeah that can happen when plotly pushes updates. Did you change the required plotly version by chance?

**Comment by computron** [comment]
I added a skip to the offending test and added an issue about it: #348 

@Qi-max please fix the remaining items on the code review, then I will look at it again to potentially merge. Note that when you write code that you feel is going to be confusing for others, do one of the following:
- rewrite that code to not be confusing (preferred, but not always possible when you are integrating with someone else's library)
- add a comment to explain what it is you are doing and why something was written the way it was, e.g. why there is some weird placeholder string (you can *always* do this)

File: matminer/featurizers/structure.py [hunk 1]
```diff
@@ -2208,6 +2228,561 @@ def implementors(self):
...
+        self.dataset = CIFDataWrapper(X, y, **self._dataset_kwargs)
+        model = self._initialize_model()
+
+        # If checkpoint_interval > num_epochs, set it as num_epochs/2
+        if self.save_checkpoint and \
+                self.checkpoint_interval >= self._num_epochs:
+            self.checkpoint_interval = math.ceil(self._num_epochs / 2)
+
+        # Get the CGCNN model
+        if self.use_pretrained:
```
[the remark this answers is not in the document]
Comment (Qi-max) [reply matminer/featurizers/structure.py]: The pretrained models only have the model settings, not directly the pickled models. In setting up the models, several parameters need to be extracted from the CIFDataWrapper, please see _initialize_model for more details.

File: matminer/featurizers/structure.py [hunk 2]
```diff
@@ -2208,6 +2228,561 @@ def implementors(self):
...
+            checkpoint_interval (int):
+                Save checkpoint every n epochs. If the epochs is less than
+                this checkpoint_interval, will reset the checkpoint_interval
+                as int(epochs/2).
+            del_checkpoint (bool):
+                Whether to delete checkpoints if training ends successfully.
+            output_path (str):
+                Output path to save the model and checkpoints, etc.
+                Output path, save model and others.
+            **cgcnn_kwargs (optional): settings of CGCNN, containing:
```
[the remark this answers is not in the document]
Comment (computron) [reply matminer/featurizers/structure.py]: ok

**Pull request merged**